storage/src/mmr/hasher.rs

```diff
@@ -46,23 +82,85 @@ impl<'a, H: CHasher> Hasher<'a, H> {
         self.finalize_reset()
     }
 
-    pub(crate) fn update_with_pos(&mut self, pos: u64) {
-        self.hasher.update(&pos.to_be_bytes());
+    fn hash(&mut self, data: &[u8]) -> H::Digest {
+        self.hasher.update(data);
+        self.finalize_reset()
     }
-    pub(crate) fn update_with_hash(&mut self, hash: &H::Digest) {
-        self.hasher.update(hash.as_ref());
+}
+
+/// Hasher the MMR uses for computing leaf, node and root hashes when the tree is being grafted onto
+/// another MMR.  If base_mmr is `None`, it behaves like a normal `Basic` hasher, and otherwise the
+/// leaf hash computation incorporates the hash of the node from the base tree onto which this leaf
+/// is grafted.
+pub struct Grafting<H: CHasher, S: Storage<H::Digest>> {
+    hasher: Basic<H>,
+    height: u32,
+    base_mmr: Option<Box<S>>,
+}
+
+impl<H: CHasher, S: Storage<H::Digest>> Grafting<H, S> {
+    pub fn new(hasher: Basic<H>, height: u32) -> Self {
+        Self {
+            hasher,
+            height,
+            base_mmr: None,
+        }
     }
-    pub(crate) fn update_with_element(&mut self, element: &[u8]) {
-        self.hasher.update(element);
+
+    pub fn basic(&mut self) -> &mut Basic<H> {
+        &mut self.hasher
     }
-    pub(crate) fn finalize_reset(&mut self) -> H::Digest {
-        self.hasher.finalize()
+
+    /// Turns on grafting mode, which allows the MMR to be grafted onto another MMR.
+    pub fn graft(&mut self, base_mmr: Box<S>) {
```

Comment:
nit: it seems weird you can create a `Grafting` without providing a `base_mmr` in `new`

Reply:
this is no longer the case